### PR TITLE
feature: add cloudflare @effect/sql-d1 package

### LINKED ISSUE #2700: From Discord: @effect/sql with Cloudflare D1
# Summary
The conversation revolves around an attempt to use `@effect/sql` with Cloudflare's D1 (SQLite) database. Here are the key points and takeaways:

1. **Initial Inquiry**: The user, phischer, inquired about the compatibility of `@effect/sql` with Cloudflare D1, wondering if it's possible to use them together given the current clients.

2. **Assumption**: Another user, datner_, suggested that if `@effect/sql` works with `better-sqlite3`, it should theoretically work with Cloudflare D1, as there might not be a significant difference in how they operate at a basic level.

3. **Implementation Attempt**: Phischer managed to get `@effect/sql` working with Cloudflare D1 by modifying the client from `sql-sqlite-node` to work with `D1Database`. However, this solution required multiple `@ts-ignore` annotations to compile, indicating potential type mismatches or unsupported operations. Due to these uncertainties and the makeshift nature of the solution, phischer expressed a lack of trust in the stability and reliability of this approach and decided to use an alternative (drizzle) for the time being.

4. **Code Example**: Phischer shared a code snippet demonstrating how they adapted `@effect/sql` for use with Cloudflare D1. The code involves creating a SQL client that interacts with the D1 database, executing a SQL query, and handling the results.

5. **Challenges**: The main challenges highlighted include:
   - The need to adapt the client to work with a database object (`D1Database`) instead of a filename, which is what the original `@effect/sql-sqlite-node` client expects.
   - Handling the asynchronous nature of D1's query results, which differ from the synchronous "try" pattern used in the existing client.
   - The necessity to bypass TypeScript's type checking with `@ts-ignore` to make the code compile, indicating potential type safety issues.

6. **Conclusion**: While it is technically possible to use `@effect/sql` with Cloudflare D1 by adapting existing clients, the process involves significant modifications and workarounds that may compromise type safety and reliability. As such, users might consider looking for more straightforward or officially supported solutions for their database needs on Cloudflare Workers.

# Discord thread

https://discord.com/channels/795981131316985866/1236321129941958726

### PR DESCRIPTION
**Pull request by ecyrbe**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Add a cloudflare D1 sqlite client.  D1 don't support transactions directly (only batches).
Since it was asked multiple times for it to be supported

```ts

import type { D1Database } from "@cloudflare/workers-types"
import { D1Client } from "@effect/sql-d1"
import { Config, Context, Data, Effect, Layer } from "effect"
import { Miniflare } from "miniflare"

export class MiniflareError extends Data.TaggedError("MiniflareError")<{
  cause: unknown
}> {}

export class D1Miniflare extends Context.Tag("test/D1Miniflare")<
  D1Miniflare,
  Miniflare
>() {
  static Live = Layer.scoped(
    this,
    Effect.acquireRelease(
      Effect.try({
        try: () =>
          new Miniflare({
            modules: true,
            d1Databases: {
              DB: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
            },
            script: ""
          }),
        catch: (cause) => new MiniflareError({ cause })
      }),
      (miniflare) => Effect.promise(() => miniflare.dispose())
    )
  )

  static ClientLive = Layer.unwrapEffect(
    Effect.gen(function*() {
      const miniflare = yield* D1Miniflare
      const db: D1Database = yield* Effect.tryPromise(() => miniflare.getD1Database("DB"))
      return D1Client.layer({
        db: Config.succeed(db)
      })
    })
  ).pipe(Layer.provide(this.Live))
}

Effect.gen(function*(_) {
  const sql = yield* D1Client.D1Client
  yield* sql`CREATE TABLE test (id INTEGER PRIMARY KEY, name TEXT)`
  yield* sql`INSERT INTO test (name) VALUES ('hello')`
  let rows = yield* _(sql`SELECT * FROM test`)
  assert.deepStrictEqual(rows, [{ id: 1, name: "hello" }])
  yield* sql`INSERT INTO test (name) VALUES ('world')`
  rows = yield* sql`SELECT * FROM test`
  assert.deepStrictEqual(rows, [
    { id: 1, name: "hello" },
    { id: 2, name: "world" }
  ])
}).pipe(Effect.provide(D1Miniflare.ClientLive)))
```

## Tests
Tests are using miniflare to have a local D1 instance

- Closes #2700

### CODE REVIEW

**Comment by mikearnaldi** [comment]
Maybe we need a "batch" api in sql to support batch transactions?

**Comment by datner** [comment]
a batch api would just be the transaction api without injecting markers between statements, though I don't think all dbs support batching in this way. For example mssql doesn't lol

**Comment by mikearnaldi** [comment]
> a batch api would just be the transaction api without injecting markers between statements, though I don't think all dbs support batching in this way. For example mssql doesn't lol

no, a batching api would allow us to specify a number of sql statements in a single shot, every sql database that supports transactions also supports such batching api.

example:

```ts
sql.batchTx(
  sql`SELECT ...`,
  sql`INSERT ...`,
  sql`DELETE...`
)
```

the execution can send a single batch to dbs that support it or simply run a `BEGIN, ..., COMMIT`.

**Comment by mikearnaldi** [comment]
> > a batch api would just be the transaction api without injecting markers between statements, though I don't think all dbs support batching in this way. For example mssql doesn't lol
> 
> no, a batching api would allow us to specify a number of sql statements in a single shot, every sql database that supports transactions also supports such batching api.
> 
> example:
> 
> ```ts
> sql.batchTx(
>   sql`SELECT ...`,
>   sql`INSERT ...`,
>   sql`DELETE...`
> )
> ```
> 
> the execution can send a single batch to dbs that support it or simply run a `BEGIN, ..., COMMIT`.

note that this is also common of products that put http apis on top of databases, while I believe this is a bad practice and that such dbs can't be used in most of the use cases where you'd need a transaction we should probably still support it.

@tim-smart we have to decide though what to do with the `withTransaction` api, how should that behave on a db that doesn't support it? the choices are:
1) make it defect
2) make it a no-op
3) make it a no-op and log a warning

**Comment by ecyrbe** [comment]
@mikearnaldi added first suggestion for now. it defects when using transactions.

**Comment by datner** [comment]
The more user-friendly choice is no-op and warn, but it might come as a surprise to the user

**Comment by ecyrbe** [comment]
My fear is that `no-op + warning` can go undetected to production.  I prefer defect personnaly.

File: packages/sql-d1/src/D1Client.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+  /** Not supported in d1 */
+  readonly updateValues: never
+}
+
+/**
+ * @category models
+ * @since 1.0.0
+ */
+export interface BackupMetadata {
```
Comment (tim-smart) [inline packages/sql-d1/src/D1Client.ts:48]: Not used?

File: packages/sql-d1/src/D1Client.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        executeRaw(sql, params) {
+          return runRaw(sql, params)
+        },
+        executeStream(_sql, _params) {
+          return Effect.dieMessage("executeStream not implemented")
+        }
+      })
+    })
+
+    const semaphore = yield* Effect.makeSemaphore(1)
```
Comment (tim-smart) [inline packages/sql-d1/src/D1Client.ts:166]: You can remove the semaphore, as there is no transactions.

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: packages/sql-d1/src/D1Client.ts (hunk 2 above)
Comment (ecyrbe) [reply packages/sql-d1/src/D1Client.ts]: Fixed

**Comment by mikearnaldi** [comment]
> My fear is that `no-op + warning` can go undetected to production. I prefer defect personnaly.

Same here

**Review by tim-smart — APPROVED** [review]

**Comment by tim-smart** [comment]
🚀

**Pull request merged**